### PR TITLE
Add LaputaVR device driver

### PR DESCRIPTION
**Pull request by mars979**

### CODE REVIEW

File: server_src/vrpn_Generic_server_object.C [hunk 1]
```diff
@@ -5101,6 +5102,32 @@ int vrpn_Generic_Server_Object::setup_nVidia_shield_USB(char *&pch, char *line,
...
+
+  VRPN_CONFIG_NEXT();
+  int ret = sscanf(pch, "%511s", s2);
+  if (ret != 1) {
+    fprintf(stderr, "Bad Laputa line: %s\n", line);
+    return -1;
+  }
+
+  // Open the Laputa
+  if (verbose) {
```
Comment (godbyk) [inline server_src/vrpn_Generic_server_object.C]: I'd move this inside of the `#ifdef VRPN_USE_HID` block.

File: vrpn_Laputa.C [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    // Skip past the report type and num_reports bytes and
+    // start parsing there.
+    vrpn_uint8 *bufptr = &buffer[2];
+
+    // The next two bytes are an increasing counter that changes by 1 for
+    // every report.
+    vrpn_uint16 report_index;
+    report_index =
```
Comment (godbyk) [inline vrpn_Laputa.C]: Why not declare and assign on the same line?

File: vrpn_Laputa.C [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        vrpn_unbuffer_from_little_endian<vrpn_uint16, vrpn_uint8>(bufptr);
+    channel[1] = report_index;
+
+    // The next two bytes are zero, so we skip them
+    vrpn_uint16 skip;
+    skip = vrpn_unbuffer_from_little_endian<vrpn_uint16, vrpn_uint8>(bufptr);
```
Comment (godbyk) [inline vrpn_Laputa.C]: Ditto.

File: vrpn_Laputa.C [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    // The next entry is temperature, and it may be in hundredths of a degree C
+    vrpn_uint16 temperature;
+    const double temperature_scale = 0.01;
+    temperature =
```
Comment (godbyk) [inline vrpn_Laputa.C]: Ditto.

File: vrpn_Laputa.C [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        unpackVector(bufptr, gyroscope_raw);
+        bufptr += 8;
+
+        // Compute the double values using default calibration.
+        // The accelerometer data goes into analogs 0,1,2.
+        // The gyro data goes into analogs 3,4,5.
+        // The magnetomoter data goes into analogs 6,7,8.
+        const double accelerometer_scale = 0.0001;
+        const double gyroscope_scale = 0.0001;
+        channel[2] = accelerometer_raw[0] * accelerometer_scale;
```
Comment (godbyk) [inline vrpn_Laputa.C]: I always forget C++'s rules for this. Will `accelerometer_raw[0]` get converted to a double, or will it all get converted to an int and then cast back to a double (losing the fractional part in the process)?

File: server_src/vrpn_Generic_server_object.C (hunk 1 above)
Comment (mars979) [inline server_src/vrpn_Generic_server_object.C]: Ok, that's fixed

File: vrpn_Laputa.C (hunk 2 above)
Comment (mars979) [inline vrpn_Laputa.C]: Ok, that's fixed

File: vrpn_Laputa.C (hunk 3 above)
Comment (mars979) [inline vrpn_Laputa.C]: Ok, that's fixed

File: vrpn_Laputa.C [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+// USB vendor and product IDs for the models we support
+static const vrpn_uint16 LAPUTA_VENDOR = 0x2633;
+static const vrpn_uint16 LAPUTA_PRODUCT = 0x0006;
+static const vrpn_int32 NUM_CHANNELS = 11;
+
+vrpn_Laputa::vrpn_Laputa(const char *name, vrpn_Connection *c)
+    : vrpn_Analog(name, c)
+    , vrpn_HidInterface(
+          m_filter = new vrpn_HidProductAcceptor(LAPUTA_VENDOR, LAPUTA_PRODUCT))
```
Comment (rpavlik) [inline vrpn_Laputa.C]: You'll want to look at updated HID drivers - this is a deprecated way of initializing them.

File: vrpn_Laputa.C [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+    vrpn_Analog::num_channel = NUM_CHANNELS;
+
+    memset(channel, 0, sizeof(channel));
+    memset(last, 0, sizeof(last));
+}
+
+vrpn_Laputa::~vrpn_Laputa() { delete m_acceptor; }
+
+inline void unpackVector(const vrpn_uint8 raw[8], int vector[3])
```
Comment (rpavlik) [inline vrpn_Laputa.C]: This should be at least partially redundant with the `vrpn_unbuffer` functions (the ones from big endian), if I'm understanding the code right. This sort of union funkiness is varying degrees of not defined in C++ depending on what version of the standard you're using.

File: vrpn_Laputa.C (hunk 5 above)
Comment (rpavlik) [inline vrpn_Laputa.C]: don't know, would probably be safest to `static_cast<double>()` around the `accelerometer_raw` parts

File: vrpn_Laputa.C (hunk 6 above)
Comment (mars979) [inline vrpn_Laputa.C]: That's fixed

File: vrpn_Laputa.C (hunk 5 above)
Comment (mars979) [inline vrpn_Laputa.C]: According to the specs, it should convert int to double when you do arithmetic operations. But I added `static_cast<double>()` to `accelerometer_raw` and `gyroscope_raw` parts anyways

**Comment by mars979** [comment]
Made changes that were requested by @rpavlik

File: vrpn_Laputa.C [hunk 8]
```diff
@@ -25,7 +25,8 @@ static const vrpn_int32 NUM_CHANNELS = 11;
 vrpn_Laputa::vrpn_Laputa(const char *name, vrpn_Connection *c)
     : vrpn_Analog(name, c)
     , vrpn_HidInterface(
-          m_filter = new vrpn_HidProductAcceptor(LAPUTA_VENDOR, LAPUTA_PRODUCT))
+          m_filter = new vrpn_HidProductAcceptor(LAPUTA_VENDOR, LAPUTA_PRODUCT),
```
Comment (rpavlik) [inline vrpn_Laputa.C]: Took me a while to find a HID device that wasn't part of a family of related devices all subclassing a base - turns out the HDK is a decent example here :D  So, what you want to do is just pass the newly-created acceptor here, and don't have an `m_filter` member yourself - in your destructor you delete `m_acceptor` (which is in `vrpn_HidInterface` base class) instead where you used to delete `m_filter`.

Comment (mars979) [inline vrpn_Laputa.C]: @rpavlik Could you check again. I think it should be fixed now.

**Comment by rpavlik** [comment]
Looks like post-merge CI flagged that the other build systems (makefile, manual VS projects) weren't updated - please update those (use a text editor on the vc project)

**Pull request merged**